File: models/layers.py
import torch
import torch.nn as nn
import torch.nn.functional as F
import math
from typing import Optional, Tuple

# Import MLA from attention and RoPE utilities from the new rope.py
from .attention import MLA_Attention
from .rope import DynamicNTKRotaryEmbedding, apply_rotary_pos_emb

class RMSNorm(nn.Module):
    def __init__(self, d_model: int, eps: float = 1e-6):
        super().__init__()
        self.eps = eps
        self.weight = nn.Parameter(torch.ones(d_model))

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        norm_x = torch.mean(x.pow(2), dim=-1, keepdim=True)
        x_normed = x * torch.rsqrt(norm_x + self.eps)
        return self.weight * x_normed

class QuantCB_FFN(nn.Module):
    def __init__(self, d_model: int, d_ff: int, dropout: float = 0.1):
        super().__init__()
        self.w_1 = nn.Linear(d_model, d_ff, bias=False)
        self.activation = nn.GELU()
        self.w_2 = nn.Linear(d_ff, d_model, bias=False)
        self.dropout = nn.Dropout(dropout)
        
        # Initialization to prevent early loss spikes
        nn.init.normal_(self.w_1.weight, std=0.02)
        nn.init.normal_(self.w_2.weight, std=0.02 / math.sqrt(2 * 6)) 

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.dropout(self.w_2(self.activation(self.w_1(x))))

class QuantCB_MoE(nn.Module):
    def __init__(self, d_model: int, d_ff: int, num_experts: int = 8, top_k: int = 2):
        super().__init__()
        self.num_experts = num_experts
        self.top_k = top_k
        self.router = nn.Linear(d_model, num_experts, bias=False)
        nn.init.normal_(self.router.weight, std=0.01)
        
        self.experts = nn.ModuleList([
            QuantCB_FFN(d_model, d_ff) for _ in range(num_experts)
        ])

    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
        batch, seq_len, d_model = x.shape
        x_flat = x.view(-1, d_model) 
        
        # 1. Get Router Scores
        router_logits = self.router(x_flat)
        weights = F.softmax(router_logits, dim=-1)
        
        # 2. Select Top-K Experts
        top_k_weights, top_k_indices = torch.topk(weights, self.top_k, dim=-1)
        
        # 3. Create Static Weight Mask (The "IREE Fix")
        expert_mask = F.one_hot(top_k_indices, num_classes=self.num_experts).float()
        
        static_weights = (top_k_weights.unsqueeze(-1) * expert_mask).sum(dim=1)
        static_weights = static_weights / (static_weights.sum(dim=-1, keepdim=True) + 1e-6)
        
        # 4. Load balancing loss
        mean_probs = weights.mean(dim=0)
        density_probs = static_weights.mean(dim=0)
        l_aux = self.num_experts * torch.sum(mean_probs * density_probs)
        
        # 5. Static Aggregation Loop
        # Explicit device and dtype binding is required to stop lifts
        final_output = torch.zeros_like(x_flat, device=x_flat.device, dtype=x_flat.dtype)
        
        for i, expert in enumerate(self.experts):
            expert_out = expert(x_flat)
            weight_i = static_weights[:, i].unsqueeze(-1)
            final_output = final_output + (expert_out * weight_i)
                
        return final_output.view(batch, seq_len, d_model), l_aux

class QuantCB_Block(nn.Module):
    def __init__(self, d_model: int, n_heads: int, d_ff: int, 
                 latent_dim: int = 128, head_dim: int = 64, 
                 num_experts: int = 8, top_k: int = 2, dropout: float = 0.1):
        super().__init__()
        self.ln_1 = RMSNorm(d_model)
        self.attn = MLA_Attention(d_model, n_heads, latent_dim, head_dim)
        self.ln_2 = RMSNorm(d_model)
        self.moe = QuantCB_MoE(d_model, d_ff, num_experts=num_experts, top_k=top_k)

    def forward(self, x: torch.Tensor, mask: Optional[torch.Tensor] = None, 
                layer_past: Optional[torch.Tensor] = None) -> Tuple[torch.Tensor, Optional[torch.Tensor], torch.Tensor]:
        
        # 1. Attention Path
        residual = x
        attn_out, present = self.attn(self.ln_1(x), mask=mask, layer_past=layer_past)
        x = residual + attn_out
        
        # 2. MoE Path
        residual = x
        moe_out, l_aux = self.moe(self.ln_2(x))
        x = residual + moe_out
        
        return x, present, l_aux

File: quantcb_rust/src/layers.rs
use candle_core::{Result, Tensor, D};
use candle_nn::{linear_no_bias, rms_norm, Linear, Module, RmsNorm, VarBuilder};

use crate::attention::MlaAttention;

/// QuantCB_FFN: Standard MLP with GELU
pub struct QuantCbFfn {
    w_1: Linear,
    w_2: Linear,
}

impl QuantCbFfn {
    pub fn new(d_model: usize, d_ff: usize, vb: VarBuilder) -> Result<Self> {
        // In Candle, standard practice is to handle initialization via the VarBuilder 
        // prior to passing it in, but the default init works well for general usage.
        let w_1 = linear_no_bias(d_model, d_ff, vb.pp("w_1"))?;
        let w_2 = linear_no_bias(d_ff, d_model, vb.pp("w_2"))?;

        Ok(Self { w_1, w_2 })
    }

    pub fn forward(&mut self, x: &Tensor) -> Result<Tensor> {
        // Dropout is omitted for inference. 
        // x -> w_1 -> gelu -> w_2
        let x = self.w_1.forward(x)?;
        let x = x.gelu()?;
        self.w_2.forward(&x)
    }
}

/// QuantCB_MoE: Mixture of Experts with Load Balancing Loss
pub struct QuantCbMoe {
    num_experts: usize,
    top_k: usize,
    router: Linear,
    experts: Vec<QuantCbFfn>,
}

impl QuantCbMoe {
    pub fn new(
        d_model: usize,
        d_ff: usize,
        num_experts: usize,
        top_k: usize,
        vb: VarBuilder,
    ) -> Result<Self> {
        let router = linear_no_bias(d_model, num_experts, vb.pp("router"))?;
        
        let mut experts = Vec::with_capacity(num_experts);
        for i in 0..num_experts {
            experts.push(QuantCbFfn::new(d_model, d_ff, vb.pp(format!("experts.{}", i)))?);
        }

        Ok(Self {
            num_experts,
            top_k,
            router,
            experts,
        })
    }

    pub fn forward(&mut self, x: &Tensor) -> Result<(Tensor, Tensor)> {
        let (batch, seq_len, d_model) = x.dims3()?;
        let num_tokens = batch * seq_len;
        let x_flat = x.reshape((num_tokens, d_model))?;

        // 1. Router logits & probabilities
        let router_logits = self.router.forward(&x_flat)?;
        let weights = candle_nn::ops::softmax(&router_logits, D::Minus1)?;

        // 2. Select Top-K Experts
        let sorted_indices = weights.arg_sort_last_dim(false)?;
        let top_k_indices = sorted_indices.narrow(D::Minus1, 0, self.top_k)?;
        
        // Gather and normalize weights
        let top_k_weights = weights.gather(&top_k_indices, D::Minus1)?;
        let sum_weights = top_k_weights.sum_keepdim(D::Minus1)?;
        
        // Add 1e-6 to denominator as in the Python code
        let eps = Tensor::new(1e-6f32, x.device())?.to_dtype(x.dtype())?;
        let sum_weights_eps = sum_weights.broadcast_add(&eps)?;
        let top_k_weights = top_k_weights.broadcast_div(&sum_weights_eps)?;

        // Extract indices to CPU for routing & loss calculation
        let top_k_indices_cpu = top_k_indices.to_vec2::<u32>()?;

        // 3. Load Balancing Loss (l_aux)
        // mean_probs: (num_experts,)
        let mean_probs = weights.mean(0)?; 
        
        // Compute density_probs efficiently on CPU without massive one_hot tensors
        let mut expert_counts = vec![0.0f32; self.num_experts];
        let total_topk = (num_tokens * self.top_k) as f32;
        for ranks in top_k_indices_cpu.iter() {
            for &exp_idx in ranks.iter() {
                expert_counts[exp_idx as usize] += 1.0;
            }
        }
        
        let density_probs_vec: Vec<f32> = expert_counts.into_iter().map(|c| c / total_topk).collect();
        let density_probs = Tensor::from_vec(density_probs_vec, (self.num_experts,), x.device())?
            .to_dtype(x.dtype())?;
        
        // l_aux = num_experts * sum(mean_probs * density_probs)
        let l_aux_val = (mean_probs.mul(&density_probs)?.sum_all()? * (self.num_experts as f64))?;

        // 4. Dispatch and Aggregate
        let mut out = x_flat.zeros_like()?;
        let flat_weights = top_k_weights.flatten_all()?;

        for i in 0..self.num_experts {
            let mut tokens_for_expert = Vec::new();
            let mut ranks_for_expert = Vec::new();

            for (token_idx, ranks) in top_k_indices_cpu.iter().enumerate() {
                for (rank, &exp_idx) in ranks.iter().enumerate() {
                    if exp_idx as usize == i {
                        tokens_for_expert.push(token_idx as u32);
                        ranks_for_expert.push(rank as u32);
                    }
                }
            }

            if tokens_for_expert.is_empty() {
                continue;
            }

            let num_exp_tokens = tokens_for_expert.len();
            let token_idx_tensor = Tensor::from_vec(tokens_for_expert.clone(), num_exp_tokens, x.device())?;

            let expert_input = x_flat.index_select(&token_idx_tensor, 0)?;
            let expert_out = self.experts[i].forward(&expert_input)?;

            let mut flat_indices = Vec::with_capacity(num_exp_tokens);
            for (t_idx, r_idx) in tokens_for_expert.iter().zip(ranks_for_expert.iter()) {
                flat_indices.push(t_idx * (self.top_k as u32) + r_idx);
            }
            let flat_idx_tensor = Tensor::from_vec(flat_indices, num_exp_tokens, x.device())?;
            
            let extracted_weights = flat_weights
                .index_select(&flat_idx_tensor, 0)?
                .unsqueeze(1)?;

            let weighted_out = expert_out.broadcast_mul(&extracted_weights)?;
            out = out.index_add(&token_idx_tensor, &weighted_out, 0)?;
        }

        let out_reshaped = out.reshape((batch, seq_len, d_model))?;
        Ok((out_reshaped, l_aux_val))
    }
}

/// QuantCB_Block: Transformer block combining Attention and MoE
pub struct QuantCbBlock {
    ln_1: RmsNorm,
    attn: MlaAttention,
    ln_2: RmsNorm,
    moe: QuantCbMoe,
}

impl QuantCbBlock {
    pub fn new(
        d_model: usize,
        n_heads: usize,
        d_ff: usize,
        latent_dim: usize,
        head_dim: usize,
        num_experts: usize,
        top_k: usize,
        vb: VarBuilder,
    ) -> Result<Self> {
        let ln_1 = rms_norm(d_model, 1e-6, vb.pp("ln_1"))?;
        let attn = MlaAttention::new(d_model, n_heads, latent_dim, head_dim, vb.pp("attn"))?;
        let ln_2 = rms_norm(d_model, 1e-6, vb.pp("ln_2"))?;
        let moe = QuantCbMoe::new(d_model, d_ff, num_experts, top_k, vb.pp("moe"))?;

        Ok(Self {
            ln_1,
            attn,
            ln_2,
            moe,
        })
    }

    pub fn forward(
        &mut self,
        x: &Tensor,
        mask: Option<&Tensor>,
        layer_past: Option<&Tensor>,
    ) -> Result<(Tensor, Option<Tensor>, Tensor)> {
        // 1. Attention Path
        let residual = x;
        let normed_x = self.ln_1.forward(x)?;
        
        let (attn_out, present) = self.attn.forward(&normed_x, mask, layer_past)?;
        let x_after_attn = residual.add(&attn_out)?;

        // 2. MoE Path
        let residual_moe = &x_after_attn;
        let normed_x_moe = self.ln_2.forward(&x_after_attn)?;
        
        let (moe_out, l_aux) = self.moe.forward(&normed_x_moe)?;
        let x_final = residual_moe.add(&moe_out)?;

        Ok((x_final, Some(present), l_aux))
    }
}